Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean

```diff
@@ -227,3 +232,492 @@ theorem Submodule.finrank_le_finrank_of_le {s t : Submodule R M} [Module.Finite
 #align submodule.finrank_le_finrank_of_le Submodule.finrank_le_finrank_of_le
 
 end
+
+
+open Cardinal Submodule Module Function FiniteDimensional
+
+attribute [local instance] nontrivial_of_invariantBasisNumber
+
+variable {R} {V : Type v}
+
+open IsNoetherian
+
+section DivisionRing
+
+variable [AddCommGroup V] {V₂ : Type v'} [AddCommGroup V₂]
+variable [Ring R] [StrongRankCondition R] [Module R V] [Module.Free R V]
+
+/-- See `FiniteDimensional.rank_lt_aleph0` for the inverse direction without `Module.Free R V`. -/
+lemma Module.rank_lt_alpeh0_iff :
+    Module.rank R V < ℵ₀ ↔ Module.Finite R V := by
+  rw [Free.rank_eq_card_chooseBasisIndex, mk_lt_aleph0_iff]
+  exact ⟨fun h ↦ Finite.of_basis (Free.chooseBasis R V),
+    fun I ↦ Finite.of_fintype (Free.ChooseBasisIndex R V)⟩
+
+theorem FiniteDimensional.finrank_of_not_finite
+    (h : ¬Module.Finite R V) :
+    finrank R V = 0 :=
+  dif_neg (rank_lt_alpeh0_iff.not.mpr h)
+
+theorem Module.finite_of_finrank_pos (h : 0 < finrank R V) :
+    Module.Finite R V := by
+  contrapose h
+  simp [finrank_of_not_finite h]
+
+theorem Module.finite_of_finrank_eq_succ {n : ℕ}
+    (hn : finrank R V = n.succ) : Module.Finite R V :=
+  Module.finite_of_finrank_pos <| by rw [hn]; exact n.succ_pos
+
+theorem Module.finite_iff_of_rank_eq_nsmul {W} [AddCommGroup W]
+    [Module R W] [Module.Free R W] {n : ℕ} (hn : n ≠ 0)
+    (hVW : Module.rank R V = n • Module.rank R W) :
+    Module.Finite R V ↔ Module.Finite R W := by
+  simp only [← rank_lt_alpeh0_iff, hVW, nsmul_lt_aleph0_iff_of_ne_zero hn]
+
+/-- If a free module is of finite rank, then the cardinality of any basis is equal to its
+`finrank`. -/
+theorem Module.mk_finrank_eq_card_basis [Module.Finite R V]
+    {ι : Type w} (h : Basis ι R V) : (finrank R V : Cardinal.{w}) = #ι := by
```

Comment:
Should this be `coe_finrank_eq_card_basis`? I guess we could also omit the `coe` ... `Cardinal.mk` doesn't appear in it.